In [1]:
# imports
import sys
import pandas as pd

In [2]:
# import custom libraries and scripts
# sys.path.append(os.path.abspath(os.path.join("../..")))
sys.path.append(".")
sys.path.append("..")

from defaults import *
from extract_dataframe import read_json
from extract_dataframe import TweetDfExtractor
from clean_tweets_dataframe import Clean_Tweets

In [3]:
# read data
global_data = pd.read_json('../data/processed_global_tweet_data.json')
global_data.head()

created_at  \
0 2022-08-07 22:31:20+00:00   
1 2022-08-07 22:31:16+00:00   
2 2022-08-07 22:31:07+00:00   
3 2022-08-07 22:31:06+00:00   
4 2022-08-07 22:31:04+00:00   

                                              source  \
0  <a href="http://twitter.com/download/android" ...   
1  <a href="http://twitter.com/download/android" ...   
2  <a href="http://twitter.com/download/android" ...   
3  <a href="http://twitter.com/download/android" ...   
4  <a href="http://twitter.com/download/iphone" r...   

                                       original_text      polarity  \
0  RT @i_ameztoy: Extra random image (I):\n\nLets... -1.250000e-01   
1  RT @IndoPac_Info: #China's media explains the ... -1.000000e-01   
2  China even cut off communication, they don't a...  0.000000e+00   
3  Putin to #XiJinping : I told you my friend, Ta...  1.000000e-01   
4  RT @ChinaUncensored: I’m sorry, I thought Taiw... -6.938894e-18   

   subjectivity lang  favorite_count  status_count  retweet_count  \
0      0.190625   en               4          8097              2   
1      0.100000   en             691          5831            201   
2      0.000000   en               0          1627              0   
3      0.350000   en               0          1627              0   
4      0.556250   en            1521         18958            381   

      screen_name original_author  followers_count  friends_count  \
0       i_ameztoy       i_ameztoy            20497           2621   
1           ZIisq           ZIisq               65            272   
2       Fin21Free       Fin21Free               85            392   
3       Fin21Free       Fin21Free               85            392   
4  VizziniDolores  VizziniDolores              910           2608   

   possibly_sensitive                                           hashtags  \
0                 NaN          [{'text': 'City', 'indices': [132, 137]}]   
1                 NaN  [{'text': 'China', 'indices': [18, 24]}, {'tex...   
2                 NaN     [{'text': 'XiJinping', 'indices': [127, 137]}]   
3                 NaN        [{'text': 'XiJinping', 'indices': [9, 19]}]   
4                 NaN                                                 []   

                                       user_mentions           place  
0  [{'screen_name': 'i_ameztoy', 'name': 'Iban Am...                  
1  [{'screen_name': 'IndoPac_Info', 'name': 'Indo...                  
2  [{'screen_name': 'ZelenskyyUa', 'name': 'Волод...     Netherlands  
3                                                 []     Netherlands  
4  [{'screen_name': 'ChinaUncensored', 'name': 'C...  Ayent, Schweiz

In [4]:
global_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22000 entries, 0 to 21999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   created_at          22000 non-null  datetime64[ns, UTC]
 1   source              22000 non-null  object             
 2   original_text       22000 non-null  object             
 3   polarity            22000 non-null  float64            
 4   subjectivity        22000 non-null  float64            
 5   lang                22000 non-null  object             
 6   favorite_count      22000 non-null  int64              
 7   status_count        22000 non-null  int64              
 8   retweet_count       22000 non-null  int64              
 9   screen_name         22000 non-null  object             
 10  original_author     22000 non-null  object             
 11  followers_count     22000 non-null  int64              
 12  friends_count       22000 non-nu

In [5]:
global_data.describe()

polarity  subjectivity  favorite_count  status_count  \
count  22000.000000  22000.000000    22000.000000  2.200000e+04   
mean       0.061325      0.283839      852.137318  5.446036e+04   
std        0.223701      0.290963     3106.077645  1.454120e+05   
min       -1.000000      0.000000        0.000000  1.000000e+00   
25%        0.000000      0.000000        2.000000  2.105750e+03   
50%        0.000000      0.200000      115.000000  1.038750e+04   
75%        0.133333      0.468824      655.000000  4.526150e+04   
max        1.000000      1.000000    65170.000000  4.108317e+06   

       retweet_count  followers_count  friends_count  possibly_sensitive  
count   22000.000000     2.200000e+04   22000.000000         6191.000000  
mean      176.750182     1.796764e+04    1563.114455            0.037151  
std       498.435765     3.030478e+05    4358.651264            0.189146  
min         0.000000     0.000000e+00       0.000000            0.000000  
25%         2.000000     5.700000e+01     137.000000            0.000000  
50%        38.000000     2.840000e+02     487.000000            0.000000  
75%       187.000000     1.324500e+03    1599.000000            0.000000  
max     17409.000000     1.449852e+07  208360.000000            1.000000

## Pre-process

In [6]:
cleaner = Clean_Tweets(global_data)

Automation in Action...!!!


### Remove duplicated rows

In [7]:
global_data.shape

(22000, 17)

In [8]:
global_data = cleaner.drop_duplicate(global_data)
global_data.shape

(7440, 17)

As we can see there were too many duplicates

### Extract source clearly

In [9]:
global_data['source']

0        <a href="http://twitter.com/download/android" ...
1        <a href="http://twitter.com/download/android" ...
2        <a href="http://twitter.com/download/android" ...
3        <a href="http://twitter.com/download/android" ...
4        <a href="http://twitter.com/download/iphone" r...
                               ...                        
21974    <a href="https://mobile.twitter.com" rel="nofo...
21987    <a href="http://twitter.com/download/iphone" r...
21989    <a href="http://twitter.com/download/iphone" r...
21991    <a href="http://twitter.com/download/android" ...
21997    <a href="http://twitter.com/download/android" ...
Name: source, Length: 7440, dtype: object

In [10]:
global_data['source'] = cleaner.extract_device_name(global_data)

In [11]:
global_data['source']

0        Twitter for Android
1        Twitter for Android
2        Twitter for Android
3        Twitter for Android
4         Twitter for iPhone
                ...         
21974        Twitter Web App
21987     Twitter for iPhone
21989     Twitter for iPhone
21991    Twitter for Android
21997    Twitter for Android
Name: source, Length: 7440, dtype: object

source has been clearly extracted

### Remove tweets that are not english

In [12]:
global_data.lang.value_counts()

en    7440
Name: lang, dtype: int64

All tweets are in English, there is nothing to remove

### Missing values

In [13]:
global_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7440 entries, 0 to 21997
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   created_at          7440 non-null   datetime64[ns, UTC]
 1   source              7440 non-null   object             
 2   original_text       7440 non-null   object             
 3   polarity            7440 non-null   float64            
 4   subjectivity        7440 non-null   float64            
 5   lang                7440 non-null   object             
 6   favorite_count      7440 non-null   int64              
 7   status_count        7440 non-null   int64              
 8   retweet_count       7440 non-null   int64              
 9   screen_name         7440 non-null   object             
 10  original_author     7440 non-null   object             
 11  followers_count     7440 non-null   int64              
 12  friends_count       7440 non-null

In [14]:
global_data.isna().sum()

created_at               0
source                   0
original_text            0
polarity                 0
subjectivity             0
lang                     0
favorite_count           0
status_count             0
retweet_count            0
screen_name              0
original_author          0
followers_count          0
friends_count            0
possibly_sensitive    3463
hashtags                 0
user_mentions            0
place                    0
dtype: int64

Only the feature possibility sensitive have 3, 463 records of missing value

In [15]:
global_data.possibly_sensitive.value_counts()

0.0    3866
1.0     111
Name: possibly_sensitive, dtype: int64

As we can see there are 3,866 not sensitive and 111 sensitive tweets
And only 3,977 out of 7,440 tweets are recorded for sensitivity

### Handling missing values

In [16]:
global_data = cleaner.fill_missing(df = global_data, column="possibly_sensitive", value = "unknown")

In [17]:
global_data.isna().sum()

created_at            0
source                0
original_text         0
polarity              0
subjectivity          0
lang                  0
favorite_count        0
status_count          0
retweet_count         0
screen_name           0
original_author       0
followers_count       0
friends_count         0
possibly_sensitive    0
hashtags              0
user_mentions         0
place                 0
dtype: int64

In [18]:
global_data.possibly_sensitive.value_counts()

0.0        3866
unknown    3463
1.0         111
Name: possibly_sensitive, dtype: int64

The missing values of the possibly_sensitive feature has been successfully filled

In [19]:
global_data.columns

Index(['created_at', 'source', 'original_text', 'polarity', 'subjectivity',
       'lang', 'favorite_count', 'status_count', 'retweet_count',
       'screen_name', 'original_author', 'followers_count', 'friends_count',
       'possibly_sensitive', 'hashtags', 'user_mentions', 'place'],
      dtype='object')

In [20]:
global_data[['retweet_count', 'source', 'original_text', 'hashtags', 'place']]

retweet_count               source  \
0                  2  Twitter for Android   
1                201  Twitter for Android   
2                  0  Twitter for Android   
3                  0  Twitter for Android   
4                381   Twitter for iPhone   
...              ...                  ...   
21974              3      Twitter Web App   
21987              1   Twitter for iPhone   
21989              0   Twitter for iPhone   
21991              0  Twitter for Android   
21997              0  Twitter for Android   

                                           original_text  \
0      RT @i_ameztoy: Extra random image (I):\n\nLets...   
1      RT @IndoPac_Info: #China's media explains the ...   
2      China even cut off communication, they don't a...   
3      Putin to #XiJinping : I told you my friend, Ta...   
4      RT @ChinaUncensored: I’m sorry, I thought Taiw...   
...                                                  ...   
21974  RT @metesohtaoglu: 📌📸 Map of #China's possible...   
21987  RT @NEVERBOW: China is doing #exactly what #Ru...   
21989  Minister Wu is crystal clear in his @BBCNews i...   
21991  Reports say that #China is planning to seize #...   
21997                @Reuters Thanks #Pelosi smart move.   

                                                hashtags             place  
0              [{'text': 'City', 'indices': [132, 137]}]                    
1      [{'text': 'China', 'indices': [18, 24]}, {'tex...                    
2         [{'text': 'XiJinping', 'indices': [127, 137]}]       Netherlands  
3            [{'text': 'XiJinping', 'indices': [9, 19]}]       Netherlands  
4                                                     []    Ayent, Schweiz  
...                                                  ...               ...  
21974  [{'text': 'China', 'indices': [29, 35]}, {'tex...       Seattle, WA  
21987  [{'text': 'exactly', 'indices': [29, 37]}, {'t...                    
21989  [{'text': 'Taiwan', 'indices': [168, 175]}, {'...   Toronto, Canada  
21991  [{'text': 'China', 'indices': [17, 23]}, {'tex...                    
21997          [{'text': 'Pelosi', 'indices': [16, 23]}]  🇺🇲🇷🇺🇺🇦🇫🇷🇦🇪🇮🇱🏳️‍🌈  

[7440 rows x 5 columns]

There are empty place values

In [21]:
global_data['place'].value_counts()

                                  2805
Việt Nam                           116
India                              107
United States                       72
Turn on 🔔                           57
                                  ... 
New York, New York                   1
Fontaines-Saint-Martin, France       1
🇺🇲🇺🇲🇺🇲                               1
Lisbon                               1
🇺🇲🇷🇺🇺🇦🇫🇷🇦🇪🇮🇱🏳️‍🌈                     1
Name: place, Length: 1809, dtype: int64

### Replace empty strings

In [22]:
global_data = cleaner.replace_empty_string(df = global_data, column="place", value = "unknown")
global_data['place'].value_counts()

unknown                           2805
Việt Nam                           116
India                              107
United States                       72
Turn on 🔔                           57
                                  ... 
New York, New York                   1
Fontaines-Saint-Martin, France       1
🇺🇲🇺🇲🇺🇲                               1
Lisbon                               1
🇺🇲🇷🇺🇺🇦🇫🇷🇦🇪🇮🇱🏳️‍🌈                     1
Name: place, Length: 1809, dtype: int64

The empty strings in the place (location) feature has been successfully filled

### Convert to datetime

In [23]:
global_data = cleaner.convert_to_datetime(global_data)
global_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7440 entries, 0 to 21997
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   created_at          7440 non-null   datetime64[ns, UTC]
 1   source              7440 non-null   object             
 2   original_text       7440 non-null   object             
 3   polarity            7440 non-null   float64            
 4   subjectivity        7440 non-null   float64            
 5   lang                7440 non-null   object             
 6   favorite_count      7440 non-null   int64              
 7   status_count        7440 non-null   int64              
 8   retweet_count       7440 non-null   int64              
 9   screen_name         7440 non-null   object             
 10  original_author     7440 non-null   object             
 11  followers_count     7440 non-null   int64              
 12  friends_count       7440 non-null

In [24]:
global_data.head()

created_at               source  \
0 2022-08-07 22:31:20+00:00  Twitter for Android   
1 2022-08-07 22:31:16+00:00  Twitter for Android   
2 2022-08-07 22:31:07+00:00  Twitter for Android   
3 2022-08-07 22:31:06+00:00  Twitter for Android   
4 2022-08-07 22:31:04+00:00   Twitter for iPhone   

                                       original_text      polarity  \
0  RT @i_ameztoy: Extra random image (I):\n\nLets... -1.250000e-01   
1  RT @IndoPac_Info: #China's media explains the ... -1.000000e-01   
2  China even cut off communication, they don't a...  0.000000e+00   
3  Putin to #XiJinping : I told you my friend, Ta...  1.000000e-01   
4  RT @ChinaUncensored: I’m sorry, I thought Taiw... -6.938894e-18   

   subjectivity lang  favorite_count  status_count  retweet_count  \
0      0.190625   en               4          8097              2   
1      0.100000   en             691          5831            201   
2      0.000000   en               0          1627              0   
3      0.350000   en               0          1627              0   
4      0.556250   en            1521         18958            381   

      screen_name original_author  followers_count  friends_count  \
0       i_ameztoy       i_ameztoy            20497           2621   
1           ZIisq           ZIisq               65            272   
2       Fin21Free       Fin21Free               85            392   
3       Fin21Free       Fin21Free               85            392   
4  VizziniDolores  VizziniDolores              910           2608   

  possibly_sensitive                                           hashtags  \
0            unknown          [{'text': 'City', 'indices': [132, 137]}]   
1            unknown  [{'text': 'China', 'indices': [18, 24]}, {'tex...   
2            unknown     [{'text': 'XiJinping', 'indices': [127, 137]}]   
3            unknown        [{'text': 'XiJinping', 'indices': [9, 19]}]   
4            unknown                                                 []   

                                       user_mentions           place  
0  [{'screen_name': 'i_ameztoy', 'name': 'Iban Am...         unknown  
1  [{'screen_name': 'IndoPac_Info', 'name': 'Indo...         unknown  
2  [{'screen_name': 'ZelenskyyUa', 'name': 'Волод...     Netherlands  
3                                                 []     Netherlands  
4  [{'screen_name': 'ChinaUncensored', 'name': 'C...  Ayent, Schweiz

### save current dataset

In [25]:
global_data.to_csv('../data/clean_data.csv', index = False)
global_data.to_json('../data/clean_data.json', indent = 4)
print('clean data saved successfully')

clean data saved successfully
